gnmi_server/gnoi_system.go

```diff
@@ -298,15 +298,93 @@ func (srv *Server) Traceroute(req *syspb.TracerouteRequest, stream syspb.System_
 	return status.Errorf(codes.Unimplemented, "Method system.Traceroute is unimplemented.")
 }
 
-// SetPackage is unimplemented.
-func (srv *Server) SetPackage(stream syspb.System_SetPackageServer) error {
-	ctx := stream.Context()
+func (srv *Server) SetPackage(rs syspb.System_SetPackageServer) error {
+	ctx := rs.Context()
+
 	_, err := authenticate(srv.config, ctx, true)
 	if err != nil {
+		log.Errorf("Authentication failed: %v", err)
+		return status.Errorf(codes.PermissionDenied, "authentication failed: %v", err)
+	}
+	log.V(1).Info("gNOI: SetPackage request received")
+
+	// Create D-Bus client
+	dbus, err := ssc.NewDbusClient()
+	if err != nil {
+		log.Errorf("Failed to create D-Bus client: %v", err)
+		return status.Errorf(codes.Internal, "failed to create D-Bus client: %v", err)
+	}
+	defer dbus.Close()
+
+	// Receive the package request
+	req, err := rs.Recv()
+	if err != nil {
+		log.Errorf("Failed to receive package request: %v", err)
+		return err
+	}
+
+	// Validate request type
+	pkg, ok := req.GetRequest().(*syspb.SetPackageRequest_Package)
+	if !ok {
+		errMsg := fmt.Sprintf("invalid request type: %T, expected SetPackageRequest_Package", req.GetRequest())
+		log.Errorf(errMsg)
+		return status.Errorf(codes.InvalidArgument, errMsg)
+	}
+
+	// A filename and a version must be provided
+	if pkg.Package.Filename == "" {
```

Comment:
>Filename [](http://example.com/codeflow?start=16&length=8)

Will you add code to check `Package.remote_download`? I understand we only want to implement remote download.